### PR TITLE
bgpv2: fix reconciliation of services with shared VIPs

### PR DESCRIPTION
**Pull request by rastislavs**
As multiple services can share the same VIP (e.g. using `lbipam.cilium.io/sharing-key`), the service reconciliation logic needs to be adapted to not withdraw a route to VIP that is still in use by some other service.

To address that, this change introduces resource reference counting by reconciling service advertisements.

This fix is specific to BGPv2 service reconciliation only.

Part of https://github.com/cilium/cilium/issues/35018 (BGPv2 only)

### CODE REVIEW

File: pkg/bgpv1/manager/reconcilerv2/paths.go
Comment (harsimran-pabla) [inline pkg/bgpv1/manager/reconcilerv2/paths.go]: Should use pathKey from afPaths map? ( not to assume NLRI.String() is the key ).

**Review by harsimran-pabla — CHANGES_REQUESTED** [review]
Thanks, very nice fix. Minor change required, otherwise lgtm 🚀

File: pkg/bgpv1/manager/reconcilerv2/paths.go
Comment (harsimran-pabla) [inline pkg/bgpv1/manager/reconcilerv2/paths.go]: nit: Should we set it to 0, prevent going it to negative.

File: pkg/bgpv1/manager/reconcilerv2/paths.go
Comment (harsimran-pabla) [inline pkg/bgpv1/manager/reconcilerv2/paths.go]: Maybe add a comment why we have to remove it from runningAdverts. ( I suppose it is because runningAdverts map is per service, maybe better naming of this variable will be enough as well. )

File: pkg/bgpv1/manager/reconcilerv2/paths.go
Comment (rastislavs) [reply pkg/bgpv1/manager/reconcilerv2/paths.go]: not really, as we may have more than 2 services sharing the VIP (e.g. in case of 3 svcs: when one of them is deleted we want to decrement this, not set to 0)

Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/paths.go]: My understanding was that svc is deleted and path need to be removed, we will decrement ref count at line 203 and return. 

Once we reach line 219, path is removed (line 213) and if everything is accounted for, PathReference[path] will be zero at this point.

File: pkg/bgpv1/manager/reconcilerv2/paths.go
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/paths.go]: Do we need to implement this logic inside the path reconciler? We can instead let Service reconciler to calculate the desired paths with reference counting? So that we don't need to modify path reconciler.

File: pkg/bgpv1/manager/reconcilerv2/paths.go
Comment (rastislavs) [reply pkg/bgpv1/manager/reconcilerv2/paths.go]: ah indeed - sorry, replied before taking a closer look at the exact place where the comment was applied

File: pkg/bgpv1/manager/reconcilerv2/paths.go
Comment (rastislavs) [reply pkg/bgpv1/manager/reconcilerv2/paths.go]: Actually, what I did is that I moved manipulation of runningAdverts back to the original place, where the context is more clear.

File: pkg/bgpv1/manager/reconcilerv2/paths.go
Comment (rastislavs) [reply pkg/bgpv1/manager/reconcilerv2/paths.go]: To me Paths reconciler is a bit more natural place for this: firstly because we may reuse this in other reconcilers in the future if needed. Secondly, if we moved this logic to the service reconciler before the Path reconciler is called, I think the logic would need to be a bit more complex (and potentially more error prone) than just guarding AdvertisePath / WithdrawPath calls - for example we would need to manipulate [desiredPaths / currentPaths in reconcilePaths](https://github.com/cilium/cilium/blob/4f00d67cb4a4f44c487f9308e0184815ab86334b/pkg/bgpv1/manager/reconcilerv2/service.go#L317-L318). But it's doable, WDYT about this @harsimran-pabla ?

Comment (YutaroHayakawa) [reply pkg/bgpv1/manager/reconcilerv2/paths.go]: I think what I felt weird was the Service reconciler still needs to do reference counting. If we implement it within the path reconciler, Service reconciler can just put the duplicated desired routes and path reconciler can count the reference. In the current implementation, both of the Service and Path reconcilers are doing reference counting in some way. However, I guess this is the way that the change becomes minimal tho.

Comment (rastislavs) [reply pkg/bgpv1/manager/reconcilerv2/paths.go]: Yeah, I see what you mean. Probably ideal would be if we could move this completely into the Path "reconciler" (`ReconcileAFPaths` func execution). The problem is that currently it is executed per-resource (service, pool, etc.), so we would need to change it to work with the set of all resources - I think that should possible, let me see how big change that would be.

Comment (rastislavs) [reply pkg/bgpv1/manager/reconcilerv2/paths.go]: OK, I think we should do it this way - implemented a new Path reconciler func `ReconcileResourceAFPaths` in https://github.com/cilium/cilium/pull/35166/commits/f7473bf0eb3f4f380b225646efd5269dfe4d1f63 (can squash the two commits before merging if we are good with this approach). This way refcounting is kept in the Path reconciler and we can even avoid some code duplication across reconcilers.

**Review by harsimran-pabla — APPROVED** [review]

**Review by YutaroHayakawa — APPROVED** [review]
Now it looks simple enough. Thanks!

File: pkg/bgpv1/manager/reconcilerv2/paths.go
Comment (rastislavs) [inline pkg/bgpv1/manager/reconcilerv2/paths.go]: I just realized that returning the `path` (input parameter) here will not work in some cases. We need to return the actual already advertised path, as only that one contains the UUID required for path withdrawal. Will fix this.

Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/paths.go]: Ah I missed it too, I think UUID is used when withdrawing a route. Did we had any test which caught it? 

Maybe we need to save the pointer to returned path in PathReferences.

Comment (rastislavs) [reply pkg/bgpv1/manager/reconcilerv2/paths.go]: >  Did we had any test which caught it?

nope, unfortunately the unit test is not catching this (it is using fake router), I noticed that when working on BGPv1 version of this fix

> Maybe we need to save the pointer to returned path in PathReferences.

yeah, I did that, please review `paths.go` again.

**Review by harsimran-pabla — APPROVED** [review]

**Pull request merged**
